src/proxy/http/fallback.rs

```diff
@@ -0,0 +1,336 @@
+use bytes::Buf;
+use futures::{Async, Future, Poll};
+use http;
+use hyper::body::Payload;
+use proxy;
+use svc;
+
+use std::{fmt, marker::PhantomData, mem};
+
+#[derive(Debug)]
+pub struct Error<A> {
+    error: proxy::Error,
+    fallback: Option<http::Request<A>>,
+}
+
+pub type Layer<P, F, A> = MakeFallback<svc::Builder<P>, svc::Builder<F>, A>;
+
+#[derive(Debug)]
+pub struct MakeFallback<P, F, A> {
+    primary: P,
+    fallback: F,
+    _p: PhantomData<fn(A)>,
+}
+
+pub struct Service<P, F, A> {
+    primary: P,
+    fallback: F,
+    _p: PhantomData<fn(A)>,
+}
+
+pub struct MakeFuture<P, F, A>
+where
+    P: Future,
+    F: Future,
+{
+    primary: Making<P>,
+    fallback: Making<F>,
+    _p: PhantomData<fn(A)>,
+}
+
+pub struct ResponseFuture<P, F, A>
+where
+    P: Future<Error = Error<A>>,
+    F: svc::Service<http::Request<A>>,
+{
+    fallback: F,
+    state: State<P, F::Future>,
+}
+
+pub enum Body<A, B> {
+    A(A),
+    B(B),
+}
+
+enum State<P, F> {
+    Primary(P),
+    Fallback(F),
+}
+
+enum Making<T: Future> {
+    NotReady(T),
+    Ready(T::Item),
+    Done,
+}
+
+pub fn layer<P, F, A>(primary: svc::Builder<P>, fallback: svc::Builder<F>) -> Layer<P, F, A> {
+    Layer {
+        primary,
+        fallback,
+        _p: PhantomData,
+    }
+}
+
+impl<P, F, A, M> svc::Layer<M> for Layer<P, F, A>
+where
+    P: svc::Layer<M> + Clone,
+    F: svc::Layer<M> + Clone,
+    M: Clone,
+{
+    type Service = MakeFallback<P::Service, F::Service, A>;
+
+    fn layer(&self, inner: M) -> Self::Service {
+        MakeFallback {
+            primary: self.primary.clone().service(inner.clone()),
+            fallback: self.fallback.clone().service(inner),
+            _p: PhantomData,
+        }
+    }
+}
+
+impl<P, F, A, T> svc::Service<T> for MakeFallback<P, F, A>
+where
+    P: svc::Service<T>,
+    P::Error: Into<proxy::Error>,
+    F: svc::Service<T>,
+    F::Error: Into<proxy::Error>,
+    T: Clone,
+{
+    type Response = Service<P::Response, F::Response, A>;
+    type Error = proxy::Error;
+    type Future = MakeFuture<P::Future, F::Future, A>;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        let primary_ready = self.primary.poll_ready().map_err(Into::into);
+        let fallback_ready = self.fallback.poll_ready().map_err(Into::into);
+        try_ready!(fallback_ready);
```

Comment:
So, this means that if the fallback isn't ready, we'll never attempt to use the primary... is that appropriate?

Reply:
I see.. `poll_ready` doesn't actually return a `Poll`... it should have a different name or signature.

Reply:
> So, this means that if the fallback isn't ready, we'll never attempt to use the primary... is that appropriate?

The alternative is to fail the request if we try to fall back and the fallback service isn't ready. Is that preferable?

Reply:
> I see.. `poll_ready` doesn't actually return a `Poll`... it should have a different name or signature.

I think this comment was intended as a reply to https://github.com/linkerd/linkerd2-proxy/pull/248#discussion_r283576128?

Reply:
I had imagined this working by cloneing the fallback service into the response future and then using an error response to cause fallback, at which point the inner service would be polled for readiness... I don't really have my head around the EndpointStatus stuff yet enough to compare...

Reply:
This probably isn't a hard blocker given that we'll use a router in the fallback route.. but we'll probably need to do something like my suggestion to make this a truly generic component.

Reply:
> I had imagined this working by cloneing the fallback service into the response future and then using an error response to cause fallback, 

That _is_ how this works...

> at which point the inner service would be polled for readiness... 

We could do it that way rather than polling the fallback service here...in that case, we would return the error if the fallback service isn't ready. Is that preferable over exerting backpressure here?

Reply:
> we would return the error if the fallback service isn't ready.

Why? Can't the response future wait until this service is ready? It sounds like ResonseState needs another state

Reply:
Oh, wow, I hadn't thought of that but it's obvious in retrospect. Will change it around.